batch-processing/pipeline_test.go

```diff
@@ -0,0 +1,830 @@
+					Input:  "/outputs/85e42ea4f380785dd6ae5c6361399d87.jpeg",
+					Output: "/outputs/85e42ea4f380785dd6ae5c6361399d87-converted.jpeg",
+				},
+			},
+		},
+		"valid GIF": {
+			In: []*Out{
+				{
+					Url: srv.URL + "/test.gif",
+				},
+			},
+			ExpectedOutput: []*Out{
+				{
+					Url:    srv.URL + "/test.gif",
+					Input:  "/outputs/d5c423921f7654ad178b167897564b13.gif",
+					Output: "/outputs/d5c423921f7654ad178b167897564b13-converted.gif",
+				},
+			},
+		},
+		"Not an image #1": {
+			In: []*Out{
+				{
+					Url: srv.URL + "/test.txt",
+				},
+			},
+			ExpectedOutput: []*Out{
+				{
+					Url: srv.URL + "/test.txt",
+					Err: errors.New("image: unknown format"),
+				},
+			},
+		},
+		"Not an image #2 (Remote URL)": {
+			In: []*Out{
+				{
+					Url: "https://www.google.com/",
+				},
+			},
+			ExpectedOutput: []*Out{
+				{
+					Url: "https://www.google.com/",
+					Err: errors.New("image: unknown format"),
+				},
+			},
+		},
+		"invalid url": {
+			In: []*Out{
+				{
+					Url: "http://notavalidurl:8080/test.png",
+				},
+			},
+			ExpectedOutput: []*Out{
+				{
+					Url: "http://notavalidurl:8080/test.png",
+					Err: invalidHostErr(t, "http://notavalidurl:8080/test.png"),
+				},
+			},
+		},
+		"erroneous row": {
+			In: []*Out{
+				{
+					Url: "http://notavalidurl:8080/test.png",
+					Err: errors.New("wrong number of fields"),
+				},
+			},
+			ExpectedOutput: []*Out{
+				{
+					Url: "http://notavalidurl:8080/test.png",
+					Err: errors.New("wrong number of fields"),
+				},
+			},
+		},
+		"duplicates": {
+			In: []*Out{
+				{Url: srv.URL + "/test.png"},
+				{Url: srv.URL + "/test.png"},
+			},
+			ExpectedOutput: []*Out{
+				{
+					Url:    srv.URL + "/test.png",
+					Input:  "/outputs/e5765cefe1b8d33fc78315437516439d.png",
+					Output: "/outputs/e5765cefe1b8d33fc78315437516439d-converted.png",
+				},
+				{
+					Url: srv.URL + "/test.png",
+					Err: ErrDuplicateURL,
+				},
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+
+			p := NewPipeline(&Config{
+				Aws: &AWSConfig{
+					GetObject: func(input *s3.GetObjectInput) (*s3.GetObjectOutput, error) {
+						return &s3.GetObjectOutput{}, fmt.Errorf("404 not found")
+					},
+				},
+			})
+			p.maxRetries = 1
+
+			p.channels[READ] = make(chan *Out, len(test.In))
+			p.channels[DOWNLOAD] = make(chan *Out, len(test.In))
+
+			for _, v := range test.In {
+				p.channels[READ] <- v
+			}
+			close(p.channels[READ])
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			go p.Download(wg)
+			wg.Wait()
+			close(p.channels[DOWNLOAD])
+
+			var gotArr []*Out
+
+			for v := range p.channels[DOWNLOAD] {
+				gotArr = append(gotArr, v)
+			}
+
+			require.Equal(t, len(gotArr), len(test.ExpectedOutput))
+
+			for _, downloadRow := range gotArr {
+				if downloadRow.Err == nil {
+					verifyFile(t, downloadRow.Input)
+				}
+			}
+
+			require.ElementsMatch(t, gotArr, test.ExpectedOutput)
+
+			t.Cleanup(func() {
```

Comment:
You need to call `Cleanup` functions _before_ any assertions which may fail, because a failed assertion stops the execution of the test at that point. `Cleanup` is effectively deferred.

Reply:
Thanks! didn't think of failing tests :)

Reply:
This `Cleanup` call is still after assertions :)